<a href="https://colab.research.google.com/github/prog815/DLSchool-deep_learning_2018-19/blob/master/%5Boptional%5Dbash.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!apt-get moo

                 (__) 
                 (oo) 
           /------\/ 
          / |    ||   
         *  /\---/\ 
            ~~   ~~   
..."Have you mooed today?"...


В этом ноутбуке мы рассмотрим базовые команды командной строки в Linux

План
  - Команды UNIX
  - Файловая система
  - Аргументы к командам
  - Переменные окружения


# Команды в UNIX

Линукс принадлежит большому семейству операционных систем под названием UNIX, коорые происходят от ОС AT&T Unix, разроботанной еще в 1970х. Mac OS, кстати тоже ее потомок.

В операционных системах UNIX-типа для взаимодействия с ОС используют специальные команды. До появления окошек и кнопочек люди создавали файлы и запускали команды исключительно с помощью этих команд. 

Сейчас этот способ тоже исользуют, тк есть ряд преимуществ:
 - Намного более удобное подключение к компьютеру по сети
 - Иногда удобнее запустить одну команду, чем бродить по менюшкам в поисках нужной опции.
 - Ломанды можно составлять в скрипты и автоматизировать задачи
 



Небольшой ликбез по понятиям:

- **Командная строка** - вид интерфейса, когда вы набираете текстовые команды
- **Терминал** - Программа, которая отображает строку. Например: *gnome-terminal*, *xterm*, *tilda*
- **Оболочка (shell)** - Программа которая выполняет команды. Например: *bash*, *zsh*, *ssh*
- **Консоль** - Железо, например клавиатура и монитор.

Запускать команды можно как из терминала, так и из юпитера или колаба, поставив перед командой !

In [2]:
!echo "Hello World"

Hello World


# Файловая система

В Unix системах все файлы образуют структуру похожую на дерево, в каждой директории(папке) хранятся файлы и другие директории. Самая первая директория которая содержит все остальные, называется корневой (*root*)


У каждой директории есть путь (по сути, просто строка) который показывает где она находится в этом дереве. "/" - это путь корневой директории. 

---



![alt text](http://linuxcommand.org/images/file_manager.jpg)

В командной строке все команды выполняются в какой-либо директории.

Представьте, что дерево файловой системы - это лабиринт, и каждая директория - это комната с файлами и дверьми в другие директории. В каждый момент времени вы находитесь в одной комнате.

Посмотрм, как нам ходить по нашему лабиринту.

In [3]:
# Выведем путь текущей директории
!pwd

/content


In [4]:
# Посмотрим что в нашей директории есть
!ls 
#ls is for "list"

sample_data


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь. 

Это так называемые скрытые директории (hidden или dot directories/files).

In [0]:
#

In [6]:
#Создадим пустой файлик
! touch test-file
# пробел в имени - плохая идея, создадутся два файла
! touch hello world

# посмотрим что получилось
! ls

hello  sample_data  test-file  world


In [7]:
#Удаление файлов
! rm world
! ls

hello  sample_data  test-file


In [0]:
# Новая директория
! mkdir please_come_here


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь.

Это так называемые скрытые директории (hidden или dot directories/files).


In [9]:
# Покажем больше информации и отобразим скрытые файлы
!ls -al
# Первое слово - команда, а второе - аргумент

total 20
drwxr-xr-x 1 root root 4096 Sep 28 17:45 .
drwxr-xr-x 1 root root 4096 Sep 28 17:35 ..
drwxr-xr-x 1 root root 4096 Sep 18 16:07 .config
-rw-r--r-- 1 root root    0 Sep 28 17:45 hello
drwxr-xr-x 2 root root 4096 Sep 28 17:45 please_come_here
drwxr-xr-x 1 root root 4096 Aug 27 16:17 sample_data
-rw-r--r-- 1 root root    0 Sep 28 17:45 test-file




```
.  - это путь к той директории в которой мы уже находимся
.. - это путь к родительской директории
.config - это скрытая папка
```

Каждая строчка - это одна из "дверей" в другую директорию либо файл.
*Относительный* путь это не полный "адрес" директории а то, куда нам нужно пойти из места в котором мы сейчас находимся. 


```
 drwxr-xr-x 1 root root 4096 Feb 15 17:21 please_come_here
 |                           дата создания    |
 |                                         _относительный_ путь
 |
  \ эта метка значит, что это путь к другой директории
```



In [10]:
# переход в другую директорию команда cd
! pwd
! cd please_come_here
! pwd

/content
/content


Заработало?

Не должно. Тут проблема в том что юпитер запускает каждый раз новую оболочку в своей директории (той, откуда его вызвали) , и после завершения команды она закрывается.

Чтоб команды запускались в новой директории нужно поменять директорию самого юпитера.


```
%cd
```



In [11]:
! pwd
% cd please_come_here
! pwd

/content
/content/please_come_here
/content/please_come_here


Когда мы переходим в другую директорию, ее имя добавляется к нашему пути через "/"

Например, были в `/home`, перешли в директорию `Pictures `оказались в `/home/Pictures`

Абсолютный путь файла в этой директории будет /home/Picutes/image.png 

Если мы хотим взять файл из директории, которая является дочерней к той директории, где мы сейчас находимся, мы можем указать относительный путь файла.

Например, из директории `/home` мы должны обращаться к нему по пути `Pictures/image.png`

In [12]:
# команды можно комбинировать в одну символом &
! ls & ls -a & pwd
%cd ..

# Чтоб команды выполнялись одна за одной нужно использовать &&
! pwd && cd please_come_here && pwd

/content/please_come_here
.  ..
/content
/content
/content/please_come_here


## Запись и чтение в файлы

In [13]:
# Вывести текст в консоль

!echo "Sample text Пример текста"
# Echo - потому что команда выводит на выход то же, что и получает

Sample text Пример текста


In [0]:
# Вместо консоли отправить в файл

! echo "London is the capital of Great Britain" >> English.txt


In [15]:
# Прочитать файл и вывести в консоль

! cat English.txt

London is the capital of Great Britain


## Скачивание из интернета

In [16]:
# сохраняем старницу по адресу

! curl example.com >> example-com.html
! ls
! cat example-com.html


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1270  100  1270    0     0  43793      0 --:--:-- --:--:-- --:--:-- 45357
English.txt  example-com.html  hello  please_come_here	sample_data  test-file
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 50px;
        background-color: #fff;
        border-radius: 1em;
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }


In [0]:
! curl instagram.com

Иногда сервера пытаются перенаправить запрос на другой адрес. Curl не следует по таким перенаправлениям. Есть прогаммка wget которая может переходить по таким ссылкам и вообще проходить по всем ссылкам на странице и скачивать их.

In [18]:
! wget instagram.com -O insta.html

--2019-09-28 17:51:09--  http://instagram.com/
Resolving instagram.com (instagram.com)... 34.205.112.224, 52.0.99.20, 34.224.153.205, ...
Connecting to instagram.com (instagram.com)|34.205.112.224|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://instagram.com/ [following]
--2019-09-28 17:51:09--  https://instagram.com/
Connecting to instagram.com (instagram.com)|34.205.112.224|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://www.instagram.com/ [following]
--2019-09-28 17:51:09--  https://www.instagram.com/
Resolving www.instagram.com (www.instagram.com)... 157.240.3.174, 2a03:2880:f201:e5:face:b00c:0:4420
Connecting to www.instagram.com (www.instagram.com)|157.240.3.174|:443... connected.
HTTP request sent, awaiting response... 200 OK
Cookie coming from www.instagram.com attempted to set domain to i.instagram.com
Cookie coming from www.instagram.com attempted to set domain to i.instagram.

Обратите внимание на

`HTTP request sent, awaiting response... 301 Moved Permanently`

# Аргументы и параметры к командам

Обычно аргументы – это то, над чем выполняется команда. Например, команда перехода в другую директорию должна получить в качестве аргумента путь директории, куда следует перейти. Команде вывода на экран содержимого файла надо передать имя файла.

Параметры модифицируют, как бы видоизменяют работу программы. Например, с помощью команды просмотра содержимого каталога можно вывести простой список вложенных объектов. Используя же определенный параметр, можно получить подробные сведения о каждом файле и папке.

Аргументы идут через пробел как отдельные слова

Параметры идут после дефиса: ` -f` (однобуквенные) либо после двух дефисов: `--file`  

In [19]:
# календарик на весь год

!ncal 2019


                                  2019
    January           February          March             April             
Su     6 13 20 27        3 10 17 24        3 10 17 24 31     7 14 21 28   
Mo     7 14 21 28        4 11 18 25        4 11 18 25     1  8 15 22 29   
Tu  1  8 15 22 29        5 12 19 26        5 12 19 26     2  9 16 23 30   
We  2  9 16 23 30        6 13 20 27        6 13 20 27     3 10 17 24      
Th  3 10 17 24 31        7 14 21 28        7 14 21 28     4 11 18 25      
Fr  4 11 18 25        1  8 15 22        1  8 15 22 29     5 12 19 26      
Sa  5 12 19 26        2  9 16 23        2  9 16 23 30     6 13 20 27      

    May               June              July              August            
Su     5 12 19 26        2  9 16 23 30     7 14 21 28        4 11 18 25   
Mo     6 13 20 27        3 10 17 24     1  8 15 22 29        5 12 19 26   
Tu     7 14 21 28        4 11 18 25     2  9 16 23 30        6 13 20 27   
We  1  8 15 22 29        5 12 19 26     3 10 17 24 31   

In [20]:
# передаем 2 аргумента

!ncal 03 2019


    March 2019        
Su     3 10 17 24 31
Mo     4 11 18 25   
Tu     5 12 19 26   
We     6 13 20 27   
Th     7 14 21 28   
Fr  1  8 15 22 29   
Sa  2  9 16 23 30   


In [21]:
# передаем параметр -b

!ncal -b 03 2019


     March 2019       
Su Mo Tu We Th Fr Sa  
                1  2  
 3  4  5  6  7  8  9  
10 11 12 13 14 15 16  
17 18 19 20 21 22 23  
24 25 26 27 28 29 30  
31                    


In [22]:
# Однобуквенные параметры можно склеивать

!ncal -b -w 03 2019
!ncal -bw 03 2019

       March 2019          
 w| Su Mo Tu We Th Fr Sa   
 8|                 1  2   
 9|  3  4  5  6  7  8  9   
10| 10 11 12 13 14 15 16   
11| 17 18 19 20 21 22 23   
12| 24 25 26 27 28 29 30   
13| 31                     
       March 2019          
 w| Su Mo Tu We Th Fr Sa   
 8|                 1  2   
 9|  3  4  5  6  7  8  9   
10| 10 11 12 13 14 15 16   
11| 17 18 19 20 21 22 23   
12| 24 25 26 27 28 29 30   
13| 31                     


# Переменные окружения

Оболочка в себе содержит так называемые переменные окружения - глобальные константы которые доступны всем программам которые она выполняет. В них удобно хранить например путь к конфигурационным файлам или порт какого нибудь сервиса.

Переменные окружения зачастую определяют как ведет себя программа. Например, питон ищет пакеты в переменной `PYTHONPATH`

В переменной `PATH` находятся пути в которых Bash ищет исполняемые файлы для других программ


In [23]:
# обьявляем переменную и считываем ее ($ перед именем)
!FOO=kindersurprise && echo $FOO

kindersurprise


In [24]:
!echo $PATH
!echo $PYTHONPATH

/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin:/opt/bin
/env/python


Мы можем создать новую папку и заставить питон думать что в ней лежит модуль

In [25]:
! rm -r fakemodule
!mkdir fakemodule
!mkdir fakemodule/power
! echo "print('A great power comes with a great responsibility')" >> fakemodule/power/__init__.py




rm: cannot remove 'fakemodule': No such file or directory


In [26]:
!export a="Эта переменная будет доступна другим программам вызванным в той же оболочке" && echo $a
# Но тут уже не получится, для каждой команды Jupyter запускает свою оболочку 
! echo $a

Эта переменная будет доступна другим программам вызванным в той же оболочке



In [27]:
! PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule && python -c 'import power'

A great power comes with a great responsibility


конструкция `$(pwd)` подставляет вывод команды `pwd` в выражение. 


```
PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule

```
После подстановки переменных мы получим 

```
PYTHONPATH=<то что было в PYTHONPATH>:<текущая директория>/fakemodule

```
Это добавит нашу fakemodule в переменную окружения и позволит питону найти наш модуль

